kernel/arch/dreamcast/hardware/maple/maple_irq.c

```diff
@@ -18,72 +19,84 @@
 /*********************************************************************/
 /* VBlank IRQ handler */
 
+/* Maple frame used for attach/detach detection. */
+static maple_frame_t detect_frame;
+
 /* Fwd declare */
 static void vbl_autodet_callback(maple_state_t *state, maple_frame_t *frm);
 
 /* Send a DEVINFO command for the given port/unit */
-static void vbl_send_devinfo(maple_state_t *state, int p, int u) {
-    maple_device_t * dev;
-
-    dev = &state->ports[p].units[u];
-
+static bool vbl_send_devinfo(maple_frame_t *frame, int p, int u) {
     /* Reserve access; if we don't get it, forget about it */
-    if(maple_frame_lock(&dev->frame) < 0)
-        return;
+    if(maple_frame_lock(frame) < 0)
+        return false;
 
     /* Setup our autodetect frame to probe at a new device */
-    maple_frame_init(&dev->frame);
-    dev->frame.cmd = MAPLE_COMMAND_DEVINFO;
-    dev->frame.dst_port = p;
-    dev->frame.dst_unit = u;
-    dev->frame.callback = vbl_autodet_callback;
-    maple_queue_frame(&dev->frame);
+    maple_frame_init(frame);
+    frame->cmd = MAPLE_COMMAND_DEVINFO;
+    frame->dst_port = p;
+    frame->dst_unit = u;
+    frame->callback = vbl_autodet_callback;
+    maple_queue_frame(frame);
+
+    return true;
 }
 
 /* Do a potential disconnect on the named device (check to make sure it
    was connected first) */
 static void vbl_chk_disconnect(maple_state_t *state, int p, int u) {
-    if(state->ports[p].units[u].valid) {
+    (void)state;
+
+    if(maple_dev_valid(p, u)) {
 #if MAPLE_IRQ_DEBUG
         dbglog(DBG_KDEBUG, "maple: detach on device %c%c\n",
                'A' + p, '0' + u);
 #endif
 
         if(maple_driver_detach(p, u) >= 0) {
-            assert(!state->ports[p].units[u].valid);
+            assert(!maple_dev_valid(p, u));
         }
     }
 }
 
+static void vbl_chk_next_subdev(maple_state_t *state, maple_frame_t *frm, int p) {
+    maple_device_t *dev = maple_enum_dev(p, 0);
+    int u;
+
+    if (dev && dev->probe_mask) {
+        u = __builtin_ffs(dev->probe_mask);
+        dev->probe_mask &= ~(1 << (u - 1));
```

Comment:
It may be worthwhile to add a macro for this pattern, as your refactor makes it appear more than just the once.

Reply:
I don't really see what macros you would want to see here.

Reply:
Will resolve this. As described in the discord, I was mistakenly interpreting this as encoding some of the specific logic for maple. In the future it would be great to set up some repo-wide basic access macros to cover stuff (like `BIT()`) but there's no reason to require it here.